scene-lib/src/scenelib/annotations/el/LocalLocation.java

```diff
@@ -1,91 +1,217 @@
 package scenelib.annotations.el;
 
+import org.objectweb.asm.Label;
+
+import java.lang.reflect.Field;
+import java.util.Arrays;
 import java.util.Objects;
 
 /**
  * A {@link LocalLocation} holds location information for a local
- * variable: slot index, scope start, and scope length.
+ * variable: slot index, scope start, and scope end.
  */
 public final class LocalLocation {
     /**
-     * The slot index of the local variable.
+     * The start of the scopes of the element being visited.
+     * Used only for TypeReference#LOCAL_VARIABLE and TypeReference#RESOURCE_VARIABLE.
+     */
+    public final Label[] start;
+
+    /**
+     * The end of the scopes of the element being visited.
+     * Used only for TypeReference#LOCAL_VARIABLE and TypeReference#RESOURCE_VARIABLE.
      */
-    public final int index;
+    public final Label[] end;
 
     /**
-     * The start of the local variable's scope (or live range), as an offset
-     * from the beginning of the method code in bytes.
+     * The indices of the element being visited in the classfile.
+     * Used only for TypeReference#LOCAL_VARIABLE and TypeReference#RESOURCE_VARIABLE.
      */
-    public final int scopeStart;
+    public final int[] index;
 
     /**
-     * The length of the local variable's scope (or live range), in bytes.
+     * The name of the local variable being visited.
      */
-    public final int scopeLength;
+    public final String variableName;
 
-    public final String varName;
-    public final int varIndex;
+    // We now use the public ASM bytecode framework instead of our own local copy.
+    // As part of the implementation changes, the following member variables
+    // of LocalLocation no longer exist:
+    //   int scopeStart;
+    //   int scopeLength;
+    //   int varIndex;
+    // Instead, the following accessor methods must be used:
+    //   int getScopeStart();
+    //   int getScopeLength();
+    //   int getVarIndex();
 
     /**
-     * Constructs a new {@link LocalLocation}; the arguments are assigned to
-     * the fields of the same names.
+     * Construct a new LocalLocation.
```

Comment:
I think this is only for a variable that is not LOCAL_VARIABLE or RESOURCE_VARIABLE.  (But I'm not positive about that.)  Can you clarify?  If so, the `start` and `end` arrays are not used.  Can the arrays contain `null`, or can the arrays even be `null`?

Reply:
It is true that start and end are meaningless in this case.  I will add a comment to that effect.  Should getScopeStart and getScopeLength check for null labels and throw an error? Currently probably get some sort of a reference to null error.  As to the kinds of variables, I have no idea. ASM TypeReference seems to define only the two you mention, no others.

Reply:
I'm wrong, there are no null values.  Unless the user uses reflection to set the start and end label values, getScopeStart and getScopeLength will report "Labels not resolved".